tests: Add conformace tests for listenersets

davidjumani opened the pull request:
<!--  Thanks for sending a pull request! Here are some tips for you:

1. If this is your first time contributing to Gateway API, please read our
   developer guide (https://gateway-api.sigs.k8s.io/contributing/devguide/)
   and our community page (https://gateway-api.sigs.k8s.io/contributing/).
2. If this is your first time contributing to a Kubernetes project, please read
   our contributor guidelines:
   https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution
3. Please label this pull request according to what type of issue you are
   addressing, especially if this is a release targeted pull request. For
   reference on required PR/issue labels, read here:
   https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
4. If you want *faster* PR reviews, read how:
   https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it:
   https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
6. If this PR includes a new GEP please make sure you've followed the process
   outlined in our GEP overview, as this will help the community to ensure the
   best chance of positive outcomes for your proposal:
   https://gateway-api.sigs.k8s.io/geps/overview/#process
-->

**What type of PR is this?**
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind gep
/kind test

Optionally add one or more of the following kinds if applicable:
/kind deprecation
/kind failing-test
/kind flake
/kind regression
/area conformance-test
/area conformance-machinery
-->

/kind test
/area conformance-test

**What this PR does / why we need it**:

This PR contains an initial conformance test for listenersets. This aims to verify the following :
- Translation of a listenerset by an implementer
- Route inheritance of gateway-level routes by the listenerset
- Route isolation between gateways and listenersets based on parentRef and sectionName

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->
Fixes https://github.com/kubernetes-sigs/gateway-api/issues/3785

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, please enter a release note below:
-->
```release-note
Adding initial conformance tests for XListenerSets
```

davidjumani commented:
/test all

davidjumani commented:
/retest

youngnick commented on conformance/tests/listenerset-http-listener-cross-namespace.go:
```diff
@@ -0,0 +1,202 @@
+		kubernetes.NamespacesMustBeReady(t, suite.Client, suite.TimeoutConfig, []string{ns})
+
+		testCases := []http.ExpectedResponse{
+			// Requests to the route defined on the gateway (should match all allowed listeners)
```
This behavior is actually not specified in the GEP at all. Dang, we missed it.

I don't know if it's better to say:
* Routes attached directly to a Gateway attach to all Listeners from the Gateway _and all Listeners from attached ListenerSets_. (that's what this asserts)
* Routes attached directly to a Gateway attach to Listeners defined on the Gateway _specifically_, _not_ ListenerSets
* Route's MUST NOT attach to a Gateway directly if the Gateway has `allowedListeners` set. (@dprotaso was hoping to make `gateway.spec.listeners` optional to make it so that we could say "only either `spec.listeners` or `spec.allowedListeners` can be set, not both).

tbh I think this is actually a reasonable default behavior, but I think we need to ensure that we document it in the GEP if we do decide that it should be. (could be done after this PR merges if necessary).

@dprotaso, @robscott any thoughts here?

davidjumani replied:
I based this behaviour on the GEP - hence if a route is attached to a gateway, it should attach to all listener-sets on that gateway unless explicitly overwritten by another route. I think that this makes things easier for the end user as well - if they would like a route defined on the gateway that works on all listeners (eg: /health) and have individual teams manage routes on the listenersets
```
Implementations MUST treat the parent Gateways as having the merged list of all listeners from itself and attached ListenerSets
```

youngnick commented on conformance/tests/listenerset-protocol-conflict.go:
```diff
@@ -0,0 +1,146 @@
+
+var ListenerSetProtocolConflict = suite.ConformanceTest{
+	ShortName:   "ListenerSetProtocolConflict",
+	Description: "Listener Set listener with protocol conflict with a Gateway listener",
```
This one also falls under the precedence rules, so the requests to the conflicted listener should succeed, and MUST be http requests.

The conflicted Listener on the ListenerSet should be not Accepted, and the Listener on the Gateway should be marked as "Conflicted".

Maybe you could add a HTTPRoute that attaches to the conflicted Listener, and validate that traffic via that HTTPRoute should _not_ flow (maybe a different path?), and another that goes on the non-conflicted ListenerSet listener, and validates that traffic via that HTTPRoute _should_ flow.

youngnick reviewed (commented):
Thanks for this PR @davidjumani, and I'm sorry that it shows how much stuff we missed when writing the original spec!

rikatz commented:
> Thanks for this PR @davidjumani, and I'm sorry that it shows how much stuff we missed when writing the original spec!

Nick, did we updated the spec and GEP with the missing items? Should we wait for #3978  before merging this one?

rikatz commented on conformance/utils/kubernetes/helpers.go:
```diff
@@ -295,6 +324,49 @@ func GatewayMustHaveCondition(
+	waitErr := wait.PollUntilContextTimeout(
+		context.Background(),
+		1*time.Second,
+		timeoutConfig.GatewayMustHaveCondition,
```
you probably want to have its own timeout config for listenerSet?

davidjumani replied:
Good idea!

rikatz commented on conformance/utils/kubernetes/helpers.go:
```diff
@@ -295,6 +324,49 @@ func GatewayMustHaveCondition(
+		},
+	)
+
+	require.NoErrorf(t, waitErr, "error waiting for Gateway status to have a Condition matching expectations")
```
s/Gateway/ListenerSet

rikatz commented on conformance/utils/kubernetes/helpers.go:
```diff
@@ -348,55 +419,15 @@ func GatewayAndRoutesMustBeAccepted(t *testing.T, c client.Client, timeoutConfig
-			Type:   string(gatewayv1.ListenerConditionProgrammed),
-			Status: metav1.ConditionTrue,
-			Reason: "", // any reason
+	resourceRef := ResourceRef{
```
I guess you have a `func NewResourceRef` above that can be used here

rikatz commented on conformance/utils/kubernetes/helpers.go:
```diff
@@ -684,14 +749,15 @@ func parentsForRouteMatch(t *testing.T, routeName types.NamespacedName, expected
 			return false
 		}
 		if aParent.ParentRef.Name != eParent.ParentRef.Name {
+			tlog.Logf(t, "Route %s ParentReference.Name doesn't match", routeName)
```
duplicated with the line below?

rikatz commented on conformance/utils/kubernetes/helpers.go:
```diff
@@ -684,14 +749,15 @@ func parentsForRouteMatch(t *testing.T, routeName types.NamespacedName, expected
 		}
 		if !reflect.DeepEqual(aParent.ParentRef.Group, eParent.ParentRef.Group) {
-			tlog.Logf(t, "Route %s expected ParentReference.Group to be %v, got %v", routeName, eParent.ParentRef.Group, aParent.ParentRef.Group)
+			tlog.Logf(t, "Route %s expected ParentReference.Group to be %v, got %v", routeName, *eParent.ParentRef.Group, *aParent.ParentRef.Group)
```
If Group (and Kind below) are by some reason null, the whole test will panic (I have tested below).  You may want to do something like:

`tlog.Errorf(t, "something happened in %v", ptr.Deref(Xpto, gatewayv1.Kind("")))`


```
var (
	Xpto *gatewayv1.Kind
)

func TestBla(t *testing.T) {
	//Xpto = ptr.To(gatewayv1.Kind("xxxx"))
	tlog.Errorf(t, "something happened in %v", *Xpto)
}
```

rikatz commented on conformance/utils/kubernetes/helpers.go:
```diff
@@ -499,6 +530,40 @@ func GatewayListenersMustHaveConditions(t *testing.T, client client.Client, time
 
+// GatewayListenerMustHaveConditions checks if the given listener of the specified gateway has all
+// the specified conditions.
+func GatewayListenerMustHaveConditions(t *testing.T, client client.Client, timeoutConfig config.TimeoutConfig, gwName types.NamespacedName, listenerName string, conditions []metav1.Condition) {
```
I have the feeling this is very similar to `GatewayListenersMustHaveConditions`, maybe should use a different name here to disambiguate, or make one function part of the other

rikatz commented on conformance/utils/kubernetes/helpers.go:
```diff
@@ -499,6 +530,40 @@ func GatewayListenersMustHaveConditions(t *testing.T, client client.Client, time
+				if listener.Name != gatewayv1.SectionName(listenerName) {
+					continue
+				}
+				listenerFound = true
```
here is a bit of an example from above, if you have found the listener you are looking for you cannot simply exit the loop/break, as it will keep looking into conditions, right?

rikatz commented:
@davidjumani I have started for the easier part for me, on helpers, and will come back to this PR tomorrow for the rest

I appreciate the effort here! Thank you!

youngnick commented:
> Nick, did we updated the spec and GEP with the missing items? Should we wait for #3978 before merging this one?

Yes, we should add these things to #3978 before merging this one.

I'll put some comments on there with the things I think we need to add.

davidjumani commented on conformance/tests/listenerset-domain-name-conflict.go:
```diff
@@ -0,0 +1,151 @@
+				Backend:   "infra-backend-v1",
+				Namespace: ns,
+			},
+			// Requests to the listener with domain name conflict should not work
+			{
+				Request:  http.Request{Host: "conflict.com", Path: "/gateway-route"},
+				Response: http.Response{StatusCode: 404},
+			},
```
I'm confused since the GEP [mentions](https://gateway-api.sigs.k8s.io/geps/gep-1713/#listener-validation)
```
Implementations MUST treat the parent Gateways as having the merged list of all listeners from itself and attached ListenerSets and validation of this list of listeners MUST behave the same as if the list were part of a single Gateway with the relaxed listener name constraints.
```
and so I based the validation tests as though they were defined on a single gateway - My understanding was that the ordering was purely based on merging them into a gateway and not validating them any different

davidjumani replied:
Okay I see https://github.com/kubernetes-sigs/gateway-api/pull/3978 has been updated. Shall I wait for it to merge before making the changes here ?

davidjumani commented on conformance/utils/kubernetes/helpers.go:
```diff
@@ -73,6 +75,33 @@ func NewGatewayRef(nn types.NamespacedName, listenerNames ...string) GatewayRef
+// relying on a specific api version.
+type ResourceRef struct {
+	types.NamespacedName
+	ListenerNames []*gatewayv1.SectionName
```
This matches `GatewayRef`. I tried to keep it as close as possible so the majority of the code could be reused

davidjumani commented on conformance/utils/kubernetes/helpers.go:
```diff
@@ -73,6 +75,33 @@ func NewGatewayRef(nn types.NamespacedName, listenerNames ...string) GatewayRef
+	var listeners []*gatewayv1.SectionName
+
+	if len(listenerNames) == 0 {
+		listenerNames = append(listenerNames, "")
```
Good idea!

davidjumani commented on conformance/utils/kubernetes/helpers.go:
```diff
@@ -295,6 +324,49 @@ func GatewayMustHaveCondition(
+			}
+
+			if err := ConditionsHaveLatestObservedGeneration(ls, ls.Status.Conditions); err != nil {
+				return false, err
```
I've copied this over from `GatewayMustHaveCondition`. Shall I update this there as well ?

rikatz replied:
looking into other places like `HTTPRouteMustHaveCondition` I can see that calls for `ConditionsHaveLatestObservedGeneration` follow the pattern of returning nil and not an error.

I don't have a strong context here on why `GatewayMustHaveCondition` is different, but from the top of my head it may take some time for controllers to update the conditions, and if you error out on the first pass it may generate false positives.

rikatz replied:
@davidjumani thoughts here?

davidjumani commented:
Thanks @youngnick @rikatz I've addressed most of the comments and am waiting for the resolution of "HTTPRoutes attached to Gateway attach to all listeners, no matter the source" behaviour to address the rest

dprotaso commented:
/hold 

based on the comments in https://github.com/kubernetes-sigs/gateway-api/pull/3978 - there seems to be a misunderstanding of Route attachment semantics in this PR

see: https://github.com/kubernetes-sigs/gateway-api/pull/3978#issuecomment-3211055462

rikatz commented:
/assign

davidjumani commented on conformance/utils/config/timeout.go:
```diff
@@ -107,25 +116,27 @@ type TimeoutConfig struct {
+		GatewayMustHaveCondition:               180 * time.Second,
+		GatewayStatusMustHaveListeners:         60 * time.Second,
+		GatewayListenersMustHaveConditions:     60 * time.Second,
+		ListenerSetMustHaveCondition:           180 * time.Second,
```
I set `ListenerSetMustHaveCondition` to match `GatewayMustHaveCondition`
`ListenerSetListenersMustHaveConditions` matches `GatewayListenersMustHaveConditions`
Do you still think it should be reduced ?

rikatz replied:
let's see how this goes with 3 minutes, and be careful if this becomes a problem for test execution time

rikatz commented:
I still have some questions, but overall lgtm and I would like to see these tests running with other implementations. I will try to trigger it tomorrow against Istio and kgateway to see how they behave (failing is fine, as this may not be implemented yet)

I have mapped some missing tests, but not for this PR:

* ListenerSet entry name unique on same set
* count of attached routes (maybe it is ther,e but I didn't spotted it)
* Port is now optional to allow for dynamic port assignment. If the port is unspecified or set to zero, the implementation will assign a unique port.
* When a ReferenceGrant is applied to a Gateway it MUST NOT be inherited by child ListenerSets. Thus a ListenerSet listener MUST NOT access secrets granted to the Gateway listeners

rikatz commented:
/lgtm
/hold cancel

The tests looks fine and reflect what we've been discussing. Some tests will need to be updated after GEP updates are merged, but this is a great start for implementations

/cc @robscott @shaneutt

gcs278 commented on conformance/tests/listenerset-http-listener-cross-namespace.yaml:
```diff
@@ -0,0 +1,233 @@
+  - kind: XListenerSet
+    group: gateway.networking.x-k8s.io
+    name: listenerset-not-allowed
+  namespace: gateway-api-example-ns6
```
indention is off
```suggestion
    namespace: gateway-api-example-ns6
```

gcs278 commented on conformance/tests/listenerset-protocol-conflict.go:
```diff
@@ -0,0 +1,202 @@
+		ls1NN := types.NamespacedName{Name: "listenerset-with-conflict-1", Namespace: ns}
+		ls1Routes := []types.NamespacedName{
+			{Namespace: ns, Name: "listenerset-with-conflict-1-route"},
+		}
```
Same as my comment on hostname test.
  
```suggestion
		}
		kubernetes.RoutesAndParentMustBeAccepted(t, suite.Client, suite.TimeoutConfig, suite.ControllerName, listenerSetRef, &gatewayv1.HTTPRoute{}, ls1Routes...)
```

Also with ls2Routes below.

davidjumani commented:
> Are you planning to add any more tests to this PR? I saw some conversation about starting with the minimal set - that's fine with me. One thing I noticed is that we test ReferenceGrant for cross-namespace backend services, but not for cross-namespace TLS secrets. Do you think that's worth including now, or should we save it for follow-up work?

I'm hoping to submit this as is - once https://github.com/kubernetes-sigs/gateway-api/pull/4205 is merged, the tests will require changes, and new ones will need to be added. I see this as a template of the tests to come, along with getting all the necessary helper methods in early

rikatz commented on conformance/utils/kubernetes/helpers.go:
```diff
@@ -320,6 +351,76 @@ func GatewayMustHaveCondition(
+			return false, nil
+		}
+
+		if gw.Status.AttachedListenerSets == nil && count == 0 {
```
followup here (dont do it now), you could simplify with ptr.Deref(gw.Status.AttachedListenerSets, 0) and check below equality.

rikatz commented on conformance/tests/listenerset-hostname-conflict.go:
```diff
@@ -0,0 +1,200 @@
+		ls1Routes := []types.NamespacedName{
+			{Namespace: ns, Name: "listenerset-with-conflict-1-route"},
+		}
+		for _, routeNN := range ls1Routes {
```
nit: why a for on a single item array? :)

rikatz commented on conformance/tests/listenerset-http-listener-cross-namespace.yaml:
```diff
@@ -0,0 +1,233 @@
+apiVersion: gateway.networking.k8s.io/v1
```
a request as a followup: our conformance tests are getting out of control on these yaml manifests file! 

If possible, we should start adding some comments on each manifest / each resource with:
* Why the resource is being created
* What it will be used for (eg.: I can see 6 namespaces, and some cross namespace listenerset, etc)

rikatz commented:
I have finalized my review. Overall I am happy with the structure of the test and how it is being executed.

I have discussed with @davidjumani and also with other maintainers and I think we should move forward with this conformance so implementations may have something to start testing and find bugs either on their implementations or on the conformance test.

/lgtm

I will wait until @youngnick is around just to see if he has some other consideration about the approach, otherwise I will come back later this weekend to approve this.

Thanks @davidjumani , it is being a long journey :)

rikatz commented:
/approve
/hold cancel
I think we can move on with these conformances, David will be fixing some of them during the implementation of kgateway and we should also keep an eye on missing conformance.

I expect that with this merged (and updated, which should happen during this week) more implementations start moving with their conformance tests.
